### DISCUSSION TITLE
Convert Connected-component Fill color

### DISCUSSION THREAD

**Discussion by ojwebb** [ImageMagick/ImageMagick#2929] [Digital Image Processing]
Im working on Imagemagick 7 CL

Im wondering if there is a way to isolate an individual connected component and replace just the fill color of one particular polygon?

The ID of the connected-component number is 38439

`magick SP1.jpg -define connected-components:verbose=true -define connected-components:area-threshold=300  -connected-components 8 testob1.png
Objects (id: bounding-box centroid area mean-color):
  0: 3200x3200+0+0 1598.8,1599.2 9.94503e+06 srgb(255,255,255)
  241: 840x751+1184+1232 1638.8,1635.3 138588 srgb(4,4,4)
  2542: 666x576+1247+1296 1576.7,1568.2 137578 srgb(156,37,69)
  38439: 146x137+1848+1679 1929.1,1738.1 7536 srgb(251,187,27)
  46581: 138x80+1832+1768 1887.6,1803.6 6674 srgb(251,187,27)
  27776: 138x137+1707+1567 1757.6,1647.7 4439 srgb(255,255,255)`

Thanks for the help

**Comment by fmw42** [comment]
Find the id. Then repeat with only that id -define connected-components:keep_id=# and make the output binary. Use that as a mask to color the region you want.

**Reply by ojwebb** [reply]
Thanks Fred,
That was extremely helpful, i appreciate the insight. I was able to grab the appropriate solid color shape with the keep id. 
I am having a tough time grasping some of the connected components thresholding options as well. my main goal is to be able to identify if a connected component is inside of another connected component. Do you have any input on how i would be able to threshold  out adjacent or touching components?

I have been playing around with these but my geometry knowledge pretty much is limited to what i learned in high school.

connected-components:angle-threshold (from equivalent ellipse)
connected-components:circularity-threshold (4*pi*area/perimeter^2)
connected-components:diameter-threshold (sqrt(4*area/pi))
connected-components:eccentricity-threshold (from equivalent ellipse)
connected-components:major-axis-threshold (diameter from equivalent ellipse)
connected-components:minor-axis-threshold (diameter from equivalent ellipse)
connected-components:perimeter-threshold

Would any of these be useful for determine beyond the the verbose bounding box info in analyzing if they are contained?
Thanks again

**Reply by fmw42** [reply]
There is no convenient method of determining if one component is inside another directly using connected components. But you could create a binary image for each id and test the union (addition) pair wise via compare. If the union is the same as the outer region, then the other one is contained inside the outer one. Larger and smaller could be determined from either the area of the region or the area of the bounding box as a quick method.

**Comment by snibgo** [comment]
This could be complex. You could test every component against every other component. If the bounding box one either is entirely within the bounding box of the other, then one _might_be_ contained by the other; otherwise, neither contains the other.

If one _might_be_ contained by the other, fmw42's test can be used, but will find only _directly_ contained components.

For example, here the green and orange shapes are _directly_ contained within the blue. The red shape is _indirectly_ contained within the blue.
![contd](https://user-images.githubusercontent.com/33812368/101260484-37d72c00-3728-11eb-9edf-89e900df86b5.png)

**Comment by fmw42** [comment]
I believe that OpenCV can get contours that are nested. So you might consider that if these solutions with ImageMagick are not sufficient
